mmdeploy/codebase/mmcls/models/utils/attention.py

```diff
@@ -44,3 +47,169 @@ def multiheadattention__forward__ncnn(ctx, self, qkv_input):
                                      v_bias, o_weight, o_bias, self.embed_dims,
                                      self.num_heads)
     return out
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    func_name=  # noqa: E251
+    'mmcls.models.utils.attention.WindowMSA.forward',
+    extra_checkers=LibVersionChecker('mmcls', min_version='0.21.0'))
+def window_msa__forward__default(ctx, self, x, mask=None):
+    B_, N, C = x.shape
+    qkv = self.qkv(x).reshape(B_, N, 3, self.num_heads,
+                              C // self.num_heads).permute(2, 0, 3, 1, 4)
+    # replace the gather operation with the split
+    q, k, v = [i.squeeze(0) for i in torch.split(qkv, 1, 0)]
+
+    q = q * self.scale
+    attn = (q @ k.transpose(-2, -1))
+
+    relative_position_bias = self.relative_position_bias_table[
+        self.relative_position_index.view(-1)].view(
+            self.window_size[0] * self.window_size[1],
+            self.window_size[0] * self.window_size[1], -1)  # Wh*Ww,Wh*Ww,nH
+    relative_position_bias = relative_position_bias.permute(
+        2, 0, 1).contiguous()  # nH, Wh*Ww, Wh*Ww
+    attn = attn + relative_position_bias.unsqueeze(0)
+
+    if mask is not None:
+        nW = mask.shape[0]
+        attn = attn.view(B_ // nW, nW, self.num_heads, N,
+                         N) + mask.unsqueeze(1).unsqueeze(0)
+        attn = attn.view(-1, self.num_heads, N, N)
+
+    means = torch.mean(attn, self.softmax.dim, keepdim=True)[0]
```

Comment:
Why cannot we use `nn.Softmax`?

Reply:
After testing, this rewriting logic seems useless. Removed.